### ISSUE TITLE
Dictionaries no longer work in expressions

### ISSUE THREAD

**Issue by bursauxa** [gravitee-io/issues#7303]
## :collision: Describe the bug
Using dictionaries in EL no longer works, they are evaluated to `null`.

## :sunrise_over_mountains: To Reproduce
Steps to reproduce the behaviour:
1. Create manual dictionary `mydictionary`
2. Add one property `mykey -> myvalue`
3. Save, deploy
4. Create a public API `/test` with any endpoint (e.g. `http://www.google.com`)
5. Add transform-header policy to add a response header `X-Test` with value `{#dictionaries['mydictionary']['mykey']}`
6. Call the API with `curl -v` (or postman, etc.)

## :rainbow: Expected behaviour
Response should have the header `X-Test: myvalue`

## Current behaviour
Response does not have the header `X-Test`

## :movie_camera: Useful information
Other EL expressions work (tested with context attributes).

This issue is the same if the dictionary is static or dynamic, it is just easier to reproduce with a static one.

## :computer: Desktop:

Gravitee 3.15.5
Mongo storage

**Comment by mouligno** [comment]
hi @bursauxa 

thanks for reaching to us, we'll take a look at this

**Comment by mouligno** [comment]
@bursauxa i couldn't reproduce, did you see anything in the logs ?

**Comment by bursauxa** [comment]
Hello, indeed I have the following in logs:

```
09:49:11.639 [gio.sync-master1] [] ERROR i.g.g.services.sync.SyncManager - An error occurs while synchronizing dictionaries
org.springframework.data.mongodb.UncategorizedMongoDbException: Command failed with error 292 (QueryExceededMemoryLimitNoDiskUseAllowed): 'Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.' on server localhost:27017. The full response is {"ok": 0.0, "errmsg": "Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.", "code": 292, "codeName": "QueryExceededMemoryLimitNoDiskUseAllowed"}; nested exception is com.mongodb.MongoCommandException: Command failed with error 292 (QueryExceededMemoryLimitNoDiskUseAllowed): 'Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.' on server localhost:27017. The full response is {"ok": 0.0, "errmsg": "Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.", "code": 292, "codeName": "QueryExceededMemoryLimitNoDiskUseAllowed"}
	at org.springframework.data.mongodb.core.MongoExceptionTranslator.translateExceptionIfPossible(MongoExceptionTranslator.java:140)
	at org.springframework.data.mongodb.core.MongoTemplate.potentiallyConvertRuntimeException(MongoTemplate.java:2899)
	at org.springframework.data.mongodb.core.MongoTemplate.execute(MongoTemplate.java:555)
	at org.springframework.data.mongodb.core.MongoTemplate.doAggregate(MongoTemplate.java:2148)
	at org.springframework.data.mongodb.core.MongoTemplate.doAggregate(MongoTemplate.java:2117)
	at org.springframework.data.mongodb.core.MongoTemplate.aggregate(MongoTemplate.java:2111)
	at org.springframework.data.mongodb.core.MongoTemplate.aggregate(MongoTemplate.java:2012)
	at io.gravitee.repository.mongodb.management.internal.event.EventMongoRepositoryImpl.searchLatest(EventMongoRepositoryImpl.java:99)
	at jdk.internal.reflect.GeneratedMethodAccessor138.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker$RepositoryFragmentMethodInvoker.lambda$new$0(RepositoryMethodInvoker.java:289)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.doInvoke(RepositoryMethodInvoker.java:137)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.invoke(RepositoryMethodInvoker.java:121)
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:529)
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:285)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:599)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.doInvoke(QueryExecutorMethodInterceptor.java:163)
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.invoke(QueryExecutorMethodInterceptor.java:138)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:80)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy72.searchLatest(Unknown Source)
	at io.gravitee.repository.mongodb.management.MongoEventRepository.searchLatest(MongoEventRepository.java:115)
	at io.gravitee.gateway.services.sync.synchronizer.AbstractSynchronizer.lambda$searchLatestEvents$0(AbstractSynchronizer.java:98)
	at io.reactivex.internal.operators.flowable.FlowableCreate.subscribeActual(FlowableCreate.java:71)
	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
	at io.reactivex.internal.operators.flowable.FlowableFlatMapMaybe.subscribeActual(FlowableFlatMapMaybe.java:54)
	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
	at io.reactivex.Flowable.subscribe(Flowable.java:14882)
	at io.reactivex.internal.operators.parallel.ParallelFromPublisher.subscribe(ParallelFromPublisher.java:58)
	at io.reactivex.internal.operators.parallel.ParallelRunOn.subscribe(ParallelRunOn.java:70)
	at io.reactivex.internal.operators.parallel.ParallelPeek.subscribe(ParallelPeek.java:80)
	at io.reactivex.internal.operators.parallel.ParallelJoin.subscribeActual(ParallelJoin.java:58)
	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
	at io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:37)
	at io.reactivex.Flowable.subscribe(Flowable.java:14935)
	at io.reactivex.internal.operators.flowable.FlowableCountSingle.subscribeActual(FlowableCountSingle.java:34)
	at io.reactivex.Single.subscribe(Single.java:3666)
	at io.reactivex.Single.blockingGet(Single.java:2869)
	at io.gravitee.gateway.services.sync.synchronizer.DictionarySynchronizer.initialSynchronizeDictionaries(DictionarySynchronizer.java:93)
	at io.gravitee.gateway.services.sync.synchronizer.DictionarySynchronizer.synchronize(DictionarySynchronizer.java:55)
	at io.gravitee.gateway.services.sync.SyncManager.refresh(SyncManager.java:150)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.mongodb.MongoCommandException: Command failed with error 292 (QueryExceededMemoryLimitNoDiskUseAllowed): 'Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.' on server localhost:27017. The full response is {"ok": 0.0, "errmsg": "Error in $cursor stage :: caused by :: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting.", "code": 292, "codeName": "QueryExceededMemoryLimitNoDiskUseAllowed"}
	at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:175)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:359)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:280)
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:100)
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:490)
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:71)
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:255)
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:202)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:118)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:110)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:345)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:336)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommandWithConnection(CommandOperationHelper.java:222)
	at com.mongodb.internal.operation.CommandOperationHelper$5.call(CommandOperationHelper.java:208)
	at com.mongodb.internal.operation.OperationHelper.withReadConnectionSource(OperationHelper.java:582)
	at com.mongodb.internal.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:205)
	at com.mongodb.internal.operation.AggregateOperationImpl.execute(AggregateOperationImpl.java:189)
	at com.mongodb.internal.operation.AggregateOperation.execute(AggregateOperation.java:296)
	at com.mongodb.internal.operation.AggregateOperation.execute(AggregateOperation.java:41)
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:178)
	at com.mongodb.client.internal.MongoIterableImpl.execute(MongoIterableImpl.java:135)
	at com.mongodb.client.internal.MongoIterableImpl.iterator(MongoIterableImpl.java:92)
	at com.mongodb.client.internal.MongoIterableImpl.forEach(MongoIterableImpl.java:121)
	at com.mongodb.client.internal.MappingIterable.forEach(MappingIterable.java:59)
	at com.mongodb.client.internal.MappingIterable.into(MappingIterable.java:69)
	at org.springframework.data.mongodb.core.MongoTemplate.lambda$doAggregate$25(MongoTemplate.java:2191)
	at org.springframework.data.mongodb.core.MongoTemplate.execute(MongoTemplate.java:553)
	... 51 common frames omitted
```

**Comment by bursauxa** [comment]
Note that the dictionary contents are fine in MongoDB, e.g. the following query works in Mongo client:

```mongo
> db.dictionaries.findOne({name: "mydictionary"}).properties["mydictionary"]
myval
```

**Comment by bursauxa** [comment]
For information - I stopped all instances, dropped the `events` collection in Mongo, restarted all instances, redeployed the API (it was marked as undeployed), and all error logs have disappeared.

The dictionary in the API still does not work though.

**Comment by bursauxa** [comment]
Hello, after dropping the `events` collection, I also dropped the `dictionaries` collection and recreated them all, and now things work properly. I can not explain this behavior, but I guess we can live with it.

One thing of note (maybe related, maybe not) - version 3.6.0 introduced a mandatory "verb" parameter for dynamic dictionaries (https://github.com/gravitee-io/issues/issues/4746) but there was no migration script provided, so when we moved up from version 3.5, it may have created some form of inconsistent state.
